File: scala/bob/src/main/scala/Bob.scala
class Bob {
  def hey(remark: String) = {
    if (isShouting(remark))
      "Whoa, chill out!"
    else if (isQuestion(remark))
      "Sure."
    else if (isSilence(remark))
      "Fine. Be that way!"
    else
      "Whatever."
  }

  val letters = ".*[A-Z].*".r

  def isShouting(remark: String): Boolean = {
    val upcased = remark.toUpperCase
    upcased match {
      case letters() => remark == upcased
      case _         => false
    }
  }
  def isQuestion(remark: String) = remark.endsWith("?")
  def isSilence(remark: String)  = remark.trim.isEmpty
}

File: ocaml/bob/bob.ml
open Core.Std

let is_question = String.is_suffix ~suffix:"?"

let is_yelling remark =
  let has_alpha = String.exists ~f:Char.is_alpha remark in
  has_alpha && String.uppercase remark = remark

let is_silence = String.for_all ~f:Char.is_whitespace

let response_for remark =
  match remark with
  | _ when is_silence remark -> "Fine. Be that way!"
  | _ when is_yelling remark -> "Whoa, chill out!"
  | _ when is_question remark  -> "Sure."
  | _ -> "Whatever."
